### DISCUSSION TITLE
How to use Postgres Pool and attach it to context.state?

### DISCUSSION THREAD

**Discussion by halvardssm** [oakserver/oak#305] [Q&A] [answered]
Hi! I am trying to create an api with Oak and [Postgres](https://github.com/denodrivers/postgres), but I am having some troubles passing the pool around as it doesn't seem to exist when I reference it in my routes. What would be a proper way of doing such?

Here are my code snippets:
```ts
// mod.ts
import { Application, State } from "https://deno.land/x/oak@v7.3.0/mod.ts";
import { Pool } from "https://deno.land/x/postgres@v0.11.2/mod.ts";
import { router } from "./router.ts";

const pool = new Pool(config, 10);
const app = new Application({ state: { pool } });

app.use(router.routes());
app.use(router.allowedMethods());

await app.listen({ port: 8000 });
```

```ts
// router.ts
import { Router } from "https://deno.land/x/oak@v7.3.0/mod.ts";
const router = new Router();

router.post("/test", async (ctx)=>{
  const body = await ctx.request.body({ type: "json" }).value;
  const client = await ctx.state.pool.connect();

  const result = await client.queryArray`select 1`;

  client.release();

  ctx.response.body = result.rows
})

export {router}
```

**Comment by kitsonk** [comment, marked as answer]
The `ctx.state` is a clone of `ctx.app.state` that is created for each request/response interaction.

Because `.pool` is intended to be "global" and exist for the lifetime of the application, I would try to use `ctx.app.state.pool` and see if that resolves the issue.

**Reply by halvardssm** [reply]
Thanks, this works! The only problem is that it doesn't seem to get the type from the context, is this intentional?

```ts
type AppContext = RouterContext<RouteParams, State & {hash: string}>

router.post("/test", async (ctx: AppContext)=>{
  const hash = ctx.state.hash; // hash is string
  const hash = ctx.app.state.hash; // hash is any
})
```

**Reply by kitsonk** [reply]
There was some generic gymnastics that needed to be performed.  If just fixed it in 81fcefa8a52867da663a7362e4da9ed2fd43001e, so it is in `main` and I should cut a new release soon.

**Reply by halvardssm** [reply]
I see! Thanks!

**Reply by harvister** [reply]
@halvardssm did this fix your issue?  I am copying your example as is and the line

```ts
const hash = ctx.app.state.hash; // hash is any
```

is still showing as any for me.  I'm using oak 9.0.0.

I see no way to tell a Router object what the type of its app's state is

**Reply by halvardssm** [reply]
@harvister It worked when I checked it at the time, but now it is not working anymore for me either 🤔 @kitsonk did something change in the types?

**Reply by harvister** [reply]
I can create an issue for this if needed
